state/watcher.go

```diff
@@ -496,6 +497,186 @@ func watchApplicationRelations(backend modelBackend, applicationName string) Str
+	return w.outCh
+}
+
+func (w *modelMachineStartTimeWatcher) loop() error {
+	docWatcher := newCollectionWatcher(w.backend, colWCfg{col: machinesC})
+	defer func() { _ = docWatcher.Stop() }()
+
+	var (
+		timer           = w.clk.NewTimer(w.quiesceInterval)
+		timerArmed      = true
+		unprocessedDocs = make(set.Strings)
+		outCh           chan []string
+		changeSet       []string
+	)
+	defer func() { _ = timer.Stop() }()
+
+	// Collect and initial set of machine IDs; this makes the worker
+	// compatible with other workers that expect the full state to be
+	// immediately emitted once the worker starts.
+	initialSet, err := w.initialMachineSet()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	changeSet = initialSet.Values()
+	outCh = w.outCh
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case <-w.watcher.Dead():
+			return stateWatcherDeadError(w.watcher.Err())
+		case changes := <-docWatcher.Changes():
+			if len(changes) == 0 {
+				continue
+			}
+			for _, docID := range changes {
+				unprocessedDocs.Add(w.backend.docID(docID))
+			}
+
+			// Restart the timer if currently stopped.
+			if !timerArmed {
+				_ = timer.Reset(w.quiesceInterval)
+				timerArmed = true
+			}
+		case <-timer.Chan():
+			timerArmed = false
+			if len(unprocessedDocs) == 0 {
+				continue // nothing to process
+			}
+
+			changedIDs, err := w.processChanges(unprocessedDocs)
+			unprocessedDocs = make(set.Strings)
+			if err != nil {
+				return err
+			} else if changedIDs.IsEmpty() {
+				continue // nothing to report
+			}
+
+			if len(changeSet) == 0 {
+				changeSet = changedIDs.Values()
+				outCh = w.outCh
+			} else {
+				// Append new set of changes to the not yet consumed changeset
+				changeSet = append(changeSet, changedIDs.Values()...)
+			}
+		case outCh <- changeSet:
+			changeSet = changeSet[:0]
+			outCh = nil
+		}
+	}
+}
+
+func (w *modelMachineStartTimeWatcher) initialMachineSet() (set.Strings, error) {
+	coll, closer := w.db.GetCollection(machinesC)
+	defer closer()
+
+	// Select the fields we need from documents that are not referring to
+	// containers.
+	iter := coll.Find(
+		bson.D{
+			{"$or", []bson.D{
+				{{"containertype", ""}},
+				{{"containertype", bson.D{{"$exists", false}}}},
+			}},
+		},
+	).Select(bson.D{{"_id", 1}, {"life", 1}, {"agent-started-at", 1}}).Iter()
+
+	var (
+		doc modelMachineStartTimeFieldDoc
+		ids = make(set.Strings)
+	)
+	for iter.Next(&doc) {
+		id := w.backend.localID(doc.Id)
+		ids.Add(id)
+		if doc.Life != Dead {
+			w.seenDocs[id] = doc
+		}
+	}
+	return ids, iter.Close()
+}
+
+func (w *modelMachineStartTimeWatcher) processChanges(pendingDocs set.Strings) (set.Strings, error) {
+	coll, closer := w.db.GetCollection(machinesC)
+	defer closer()
+
+	// Select the fields we need from the changed documents that are not
+	// referring to containers.
+	iter := coll.Find(
+		bson.D{
+			{"_id", bson.D{{"$in", pendingDocs.Values()}}},
+			{"$or", []bson.D{
```

Comment:
this selector could also probably be put into a
notContainerQuery
or something along those lines, since it is repeated between initialSet and here